### PR TITLE
[wingman -> rllib] IMPALA MultiDiscrete changes

### PR DESCRIPTION
**Pull request by bjg2**
NOTE: This is the beginning of the pull request, so we can align. No unit tests were run, nor the changes are tested broader than we need them.

## What do these changes do?

- Enables the MultiDiscrete actions space in IMPALA.

- Adds the MultiCategorical action distribution - we use IMPALA with that distribution - out model outputs categorical logits for each action. Possibly not all action spaces work for all action distributions, tested is that Discrete works with DiagGaussian (default) and that MultiDiscrete works with categorical (MultiCategorical). Note that model actually outputs a 1-dimensional tensor, that is reshaped to logit sizes according to provided MultiDiscrete action space.

- VTrace is replaced with an implementation that works for MultiDiscrete action space as well.

### CODE REVIEW

**Comment by ericl** [comment]
Do you know what's going on with the formatting changes? It makes it kind of hard to review.

**Comment by bjg2** [comment]
> Do you know what's going on with the formatting changes? It makes it kind of hard to review.

Aligned formatting with yours impala.py, it should be ok to take a look at merge request now.

**Review by ericl — COMMENTED** [review]
Since we're changing the vtrace code now, I think it's also good to copy over the vtrace_test file from here: https://github.com/deepmind/scalable_agent/blob/master/vtrace_test.py
and at least get it to run in the normal Discrete() case. It would also be good to add a new test for the MultiDiscrete() spaces.

I would also add an end-to-end runnable script that just verifies MultiDiscrete doesn't crash with a real env (you can add the file to rllib/test/, and add an entry in multi_node_tests.sh)

File: python/ray/rllib/agents/impala/impala.py [hunk 1]
```diff
@@ -70,6 +70,9 @@
     # max number of workers to broadcast one set of weights to
     "broadcast_interval": 1,
 
+    # Actions are chosen based on this distribution, if provided
+    "dist_type": None,
```
[the remark this answers is not in the document]
Comment (bjg2) [reply python/ray/rllib/agents/impala/impala.py]: It is, as default is DiagGaussian, and we use Categorical (which is MultiCategorical for MultiDiscrete action space).

File: python/ray/rllib/agents/impala/vtrace_policy_graph.py [hunk 2]
```diff
@@ -101,6 +109,11 @@ def __init__(self,
             "Must use `truncate_episodes` batch mode with V-trace."
         self.config = config
         self.sess = tf.get_default_session()
+        self._is_discrete = False
```
[the remark this answers is not in the document]
Comment (bjg2) [reply python/ray/rllib/agents/impala/vtrace_policy_graph.py]: Agree, fixed.

File: python/ray/rllib/agents/impala/vtrace.py [hunk 3]
```diff
@@ -189,10 +190,8 @@ def from_importance_weights(log_rhos,
   `rewards` is [T, B, C], `values` is [T, B, C], `bootstrap_value` is [B, C].
 
   Args:
-    log_rhos: A float32 tensor of shape [T, B, NUM_ACTIONS] representing the
-      log importance sampling weights, i.e.
-      log(target_policy(a) / behaviour_policy(a)). V-trace performs operations
-      on rhos in log-space for numerical stability.
+    rhos: A float32 tensor of shape [T, B, NUM_ACTIONS] representing the
+      importance sampling weights, i.e. target_policy(a) / behaviour_policy(a).
```
[the remark this answers is not in the document]
Comment (pimpke) [reply python/ray/rllib/agents/impala/vtrace.py]: We were experimenting with both log prob and prob spaces, will return it to log space.

File: python/ray/rllib/agents/impala/vtrace.py [hunk 4]
```diff
@@ -33,14 +33,14 @@
 nest = tf.contrib.framework.nest
 
 VTraceFromLogitsReturns = collections.namedtuple('VTraceFromLogitsReturns', [
-    'vs', 'pg_advantages', 'log_rhos', 'behaviour_action_log_probs',
-    'target_action_log_probs'
+    'vs', 'pg_advantages', 'rhos', 'behaviour_action_policy',
+    'target_action_policy'
```
[the remark this answers is not in the document]
Comment (pimpke) [reply python/ray/rllib/agents/impala/vtrace.py]: Yup.

File: python/ray/rllib/agents/impala/vtrace.py [hunk 5]
```diff
@@ -33,14 +33,14 @@
 nest = tf.contrib.framework.nest
```
[the remark this answers is not in the document]
Comment (bjg2) [reply python/ray/rllib/agents/impala/vtrace.py]: Done.

File: python/ray/rllib/agents/impala/vtrace.py [hunk 6]
```diff
@@ -33,14 +33,14 @@
...
-    'target_action_log_probs'
+    'vs', 'pg_advantages', 'rhos', 'behaviour_action_policy',
+    'target_action_policy'
 ])
 
 VTraceReturns = collections.namedtuple('VTraceReturns', 'vs pg_advantages')
 
 
-def log_probs_from_logits_and_actions(policy_logits, actions):
+def select_policy_values_using_actions(policy_logits, actions):
```
[the remark this answers is not in the document]
Comment (bjg2) [reply python/ray/rllib/agents/impala/vtrace.py]: done.

File: python/ray/rllib/agents/impala/vtrace.py [hunk 7]
```diff
@@ -93,10 +92,10 @@ def from_logits(behaviour_policy_logits,
   NUM_ACTIONS refers to the number of actions.
 
   Args:
-    behaviour_policy_logits: A float32 tensor of shape [T, B, NUM_ACTIONS] with
+    behaviour_policy: A float32 tensor of shape [T, B, NUM_ACTIONS] with
```
[the remark this answers is not in the document]
Comment (bjg2) [reply python/ray/rllib/agents/impala/vtrace.py]: done.

**Comment by ericl** [comment]
I pushed some changes to allow vtrace_test to run on the modified implementation. Looks like most tests pass but test_vtrace_from_logits_Batch1 and test_vtrace_from_logits_Batch2 don't, could you take a look?

**Comment by bjg2** [comment]
> I pushed some changes to allow vtrace_test to run on the modified implementation. Looks like most tests pass but test_vtrace_from_logits_Batch1 and test_vtrace_from_logits_Batch2 don't, could you take a look?

Fixed those tests with https://github.com/ray-project/ray/pull/3967/commits/f161edf89e996ce734ac3aa3fd4cf079f8dfbfe9

**Review by ericl — APPROVED** [review]
LGTM (just one question still on dist_type)

**Comment by bjg2** [comment]
> I'm a bit confused, isn't it Categorical for discrete spaces?
>
> ```
>        elif isinstance(action_space, gym.spaces.Discrete):
>            return Categorical, action_space.n
> ```
> I don't think DiagGaussian ever gets used in IMPALA does it (maybe you meant APPO?)

Good catch - my bad. I was first using Box for IMPALA, as I didn't see MultiDiscrete action space at the moment. For Box, the default was DiagGaussian. When I switched to MultiDiscrete I forgot dist_type was not needed anymore. Fixed with https://github.com/ray-project/ray/pull/3967/commits/d32d253c63ca73944ada0dc887a34826e4f37b6f.

**Comment by bjg2** [comment]
Is this one waiting on something?

**Comment by ericl** [comment]
Ah looks like test_supported_spaces is failing on APPO somehow. I can take a look in a bit.

**Comment by ericl** [comment]
Merged, thanks!
